### PR TITLE
Change update mechanism in reducers for analysisFramework

### PR DESCRIPTION
**Pull request by tnagorra**
Depends on
https://github.com/toggle-corp/react-store/pull/632
https://github.com/toggle-corp/react-store/pull/674

### CODE REVIEW

File: src/redux/reducers/siloDomainData/analysisFramework.js [hunk 1]
```diff
@@ -192,11 +196,19 @@ const afViewRemoveWidget = (state, action) => {
...
+        widgetTitle,
+    } = action;
+
     const { analysisFrameworkView: { [analysisFrameworkId]: { data: analysisFramework } } } = state;
 
     const existingWidgets = analysisFramework.widgets;
-    const widgetIndex = existingWidgets.findIndex(w => getWidgetKey(w) === widget.key);
+    const widgetIndex = existingWidgets.findIndex(w => getWidgetKey(w) === widgetKey);
+
+    console.warn(widgetIndex, widgetKey);
```
Comment (AdityaKhatri) [inline src/redux/reducers/siloDomainData/analysisFramework.js]: Sir!
Sir, please!
Sir!

File: src/redux/reducers/siloDomainData/analysisFramework.js [hunk 2]
```diff
@@ -207,7 +219,14 @@ const afViewUpdateWidget = (state, action) => {
             [analysisFrameworkId]: {
                 data: {
                     widgets: {
-                        [widgetIndex]: { $merge: widget },
+                        [widgetIndex]: {
```
Comment (AdityaKhatri) [inline src/redux/reducers/siloDomainData/analysisFramework.js:220]: Sir immer?

**Review by AdityaKhatri — APPROVED** [review]

**Review by frozenhelium — APPROVED** [review]

File: src/components/general/OrientationInput/index.tsx [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+        </div>
+    ) : data.label
+);
+
+
+export default class OrientationInput extends React.PureComponent<Props, State> {
+    render() {
+        const {
+            // FIXME: string
+            label = 'Orientation',
```
Comment (tnagorra) [inline src/components/general/OrientationInput/index.tsx:40]: Please use string here

File: src/views/AnalysisFramework/WidgetEditor/index.js [hunk 4]
```diff
@@ -61,6 +61,27 @@ export default class WidgetEditor extends React.PureComponent {
     static schema = {};
     static value = {};
 
+    /*
```
Comment (tnagorra) [inline src/views/AnalysisFramework/WidgetEditor/index.js:64]: Line farming? eh?

File: src/views/AnalysisFramework/WidgetEditor/index.js [hunk 5]
```diff
@@ -74,17 +95,39 @@ export default class WidgetEditor extends React.PureComponent {
         return minSize;
     }
 
-    handleItemChange = (newWidget) => {
+    /*
```
Comment (tnagorra) [inline src/views/AnalysisFramework/WidgetEditor/index.js:98]: Line farming? eh?

File: src/views/AnalysisFramework/WidgetEditor/index.js [hunk 6]
```diff
@@ -132,57 +175,75 @@ export default class WidgetEditor extends React.PureComponent {
...
+                    {!hideButtons ? (
+                        <Fragment>
+                            <WarningButton
+                                iconName="edit"
+                                title={_ts('framework.widgetEditor', 'editTooltip')}
+                                tabIndex="-1"
+                                transparent
+                                onClick={() => this.props.onWidgetEditClick(key, widget)}
+                                disabled={widgetsDisabled}
+                                // TODO: add onClick handler
```
Comment (tnagorra) [inline src/views/AnalysisFramework/WidgetEditor/index.js:215]: Need to look into this again!

File: src/views/AnalysisFramework/WidgetEditor/index.js [hunk 7]
```diff
@@ -200,38 +261,33 @@ export default class WidgetEditor extends React.PureComponent {
...
-        const className = `
-            ${styles.content}
-            ${isDisabled ? styles.disabled : ''}
-        `;
+        const className = _cs(
+            styles.content,
+            isDisabled && styles.disabled,
+        );
+
+        /*
```
Comment (tnagorra) [inline src/views/AnalysisFramework/WidgetEditor/index.js:282]: Line farming? eh?

File: src/views/AnalysisFramework/WidgetEditor/index.js [hunk 8]
```diff
@@ -250,15 +307,10 @@ export default class WidgetEditor extends React.PureComponent {
...
-                { isDisabled && (
-                    <div
-                        className={styles.disablerMask}
-                        style={{
-                            fontSize: `${fontSize}px`,
-                        }}
-                    >
-                        { disablerMaskText }
-                    </div>
+                { isDisabled && ( // FIXME: use Message component?
```
Comment (tnagorra) [inline src/views/AnalysisFramework/WidgetEditor/index.js:310]: ALready used message component, let's remove FIXME here

File: src/views/EditEntries/Overview/index.js [hunk 9]
```diff
@@ -301,7 +301,9 @@ export default class Overview extends React.PureComponent {
                         <WidgetFaram
                             className={styles.content}
                             // NOTE: dismount on key change
-                            key={key}
+                            // NOTE: removed dismount on key change behavior
+                            // to persist active UI state
+                            // key={key}
```
Comment (tnagorra) [inline src/views/EditEntries/Overview/index.js:306]: Need to look into this further!

File: src/widgetComponents/Matrix2dInput/DimensionRow/SubdimensionRow/Cell/index.js [hunk 10]
```diff
@@ -125,7 +150,11 @@ export default class Cell extends React.PureComponent {
             disabled,
             readOnly,
             hoverStyle,
+            activeSectorKey,
```
Comment (tnagorra) [inline src/widgetComponents/Matrix2dInput/DimensionRow/SubdimensionRow/Cell/index.js:153]: Looks like the 3 props aren't used here

File: src/widgetComponents/Matrix2dInput/DimensionRow/SubdimensionRow/index.js [hunk 11]
```diff
@@ -25,16 +27,88 @@ export default class SubdimensionRow extends React.PureComponent {
...
+            subdimensionId,
+            sectorId,
+            activeSectorKey,
+        } = this.props;
+
+        if (!activeSectorKey) {
+            return false;
+        }
+
+        const subsectors = value && ((value[dimensionId] || {})[subdimensionId]);
```
Comment (tnagorra) [inline src/widgetComponents/Matrix2dInput/DimensionRow/SubdimensionRow/index.js:92]: We should use de-structuring here

File: src/widgetComponents/Matrix2dInput/DimensionRow/SubdimensionRow/index.js [hunk 12]
```diff
@@ -25,16 +27,88 @@ export default class SubdimensionRow extends React.PureComponent {
...
             ...otherProps
         } = this.props;
+
+        const isSubsectorMode = !!activeSectorKey;
+
         return {
-            sectorId: key,
+            isSubsectorMode,
+            sectorId: isSubsectorMode ? activeSectorKey : key,
+            subsectorId: isSubsectorMode ? key : undefined,
```
Comment (tnagorra) [inline src/widgetComponents/Matrix2dInput/DimensionRow/SubdimensionRow/index.js:111]: Do we also need to inject activeSectorKey here?

File: src/widgetComponents/Matrix2dInput/DimensionRow/index.js [hunk 13]
```diff
@@ -37,17 +39,20 @@ export default class DimensionRow extends React.PureComponent {
...
                 getRgbRawFromHex(rowStyle.color),
-                0.4,
+                0.3,
             ),
         );
 
+        const textBorderColor = firstColor;
         return {
+            textShadow: `-1px 0 ${textBorderColor}, 0 1px ${textBorderColor}, 1px 0 ${textBorderColor}, 0 -1px ${textBorderColor}`,
+            // textShadow: '0 0 red'
```
Comment (tnagorra) [inline src/widgetComponents/Matrix2dInput/DimensionRow/index.js:55]: Line farming? eh?

File: src/widgetComponents/Matrix2dInput/index.js [hunk 14]
```diff
@@ -50,43 +34,92 @@ export default class Matrix2dInput extends React.PureComponent {
...
+        });
+
         onChange(newValue);
     }
 
-    handleCellDrop = (dimensionId, subdimensionId, sectorId, droppedData) => {
+    handleCellDrop = (dimensionId, subdimensionId, sectorId, subsectorId, droppedData) => {
         const { type, data } = droppedData;
 
+        console.warn('Sector', sectorId, 'Subsector', subsectorId);
```
Comment (tnagorra) [inline src/widgetComponents/Matrix2dInput/index.js:113]: Console.warn sir!

File: src/widgetComponents/Matrix2dInput/index.js [hunk 15]
```diff
@@ -120,29 +177,63 @@ export default class Matrix2dInput extends React.PureComponent {
...
         } = this.props;
 
+        const { activeSectorKey } = this.state;
+
+        const headRowStyle = this.getHeadRowStyle(meta.titleRowHeight);
+        const titleColumnStyle = this.getTitleColumnStyle(meta.titleColumnWidth);
+        const subTitleColumnStyle = this.getSubTitleColumnStyle(meta.subTitleColumnWidth);
+
+        const activeSector = this.getActiveSector(sectors, activeSectorKey);
+        const subsectors = activeSector ? activeSector.subsectors : [];
```
Comment (tnagorra) [inline src/widgetComponents/Matrix2dInput/index.js:190]: We shouldn't create new array in render

File: src/widgets/edit/Matrix2d/Column/SectorContent/SubsectorRow/index.js [hunk 16]
```diff
@@ -0,0 +1,80 @@
...
+                        type="number"
+                        // FIXME: string
+                        label="Font size"
+                        className={styles.fontSizeInput}
+                        faramElementName="fontSize"
+                    />
+                    <TextInput
+                        type="number"
+                        // FIXME: string
+                        label="Width"
```
Comment (tnagorra) [inline src/widgets/edit/Matrix2d/Column/SectorContent/SubsectorRow/index.js:49]: Use string

File: src/widgets/edit/Matrix2d/Column/SectorContent/SubsectorRow/index.js [hunk 17]
```diff
@@ -0,0 +1,80 @@
...
+                    />
+                    <OrientationInput
+                        className={styles.orientationInput}
+                        faramElementName="orientation"
+                        persistantHintAndError={false}
+                    />
+                    <TextInput
+                        type="number"
+                        // FIXME: string
+                        label="Font size"
```
Comment (tnagorra) [inline src/widgets/edit/Matrix2d/Column/SectorContent/SubsectorRow/index.js:42]: Use string

File: src/widgets/edit/Matrix2d/Column/SectorContent/index.js [hunk 18]
```diff
@@ -0,0 +1,200 @@
...
+                            />
+                            <OrientationInput
+                                className={styles.orientationInput}
+                                faramElementName="orientation"
+                                persistantHintAndError={false}
+                            />
+                            <TextInput
+                                type="number"
+                                // FIXME: string
+                                label="Font size"
```
Comment (tnagorra) [inline src/widgets/edit/Matrix2d/Column/SectorContent/index.js:118]: Use string

File: src/widgets/edit/Matrix2d/Column/SectorContent/index.js [hunk 19]
```diff
@@ -0,0 +1,200 @@
...
+                                className={styles.fontSizeInput}
+                                faramElementName="fontSize"
+                            />
+                            <TextInput
+                                type="number"
+                                // FIXME: string
+                                label="Width"
```
Comment (tnagorra) [inline src/widgets/edit/Matrix2d/Column/SectorContent/index.js:125]: Use string

File: src/widgets/edit/Matrix2d/Column/index.tsx [hunk 20]
```diff
@@ -0,0 +1,130 @@
...
+const emptyList: unknown[] = [];
+const emptyObject = {};
+
+export default class Column extends React.PureComponent<Props, State> {
+    private sectorItemRendererParams = (key, elem, i) => ({
+        className: styles.sectorContent,
+        data: elem,
+        faramElementName: String(i),
+        index: i,
+        onEditButtonClick: () => {
```
Comment (tnagorra) [inline src/widgets/edit/Matrix2d/Column/index.tsx:42]: We shouldn't create new function in rendererParams

File: src/widgets/edit/Matrix2d/Column/index.tsx [hunk 21]
```diff
@@ -0,0 +1,130 @@
...
+            onGeoLinkModalVisiblityChange,
+            onLinkWidgetModalVisiblityChange,
+            onSectorContentBackButtonClick,
+            selectedSectorKey,
+            titleSelector,
+            widgetKey,
+        } = this.props;
+
+        const { sectors = emptyList } = faramValues;
+        const selectedSectorIndex = sectors
```
Comment (tnagorra) [inline src/widgets/edit/Matrix2d/Column/index.tsx:64]: We should memoize expensive actions

File: src/widgets/edit/Matrix2d/Row/DimensionContent/SubdimensionRow/index.js [hunk 22]
```diff
@@ -0,0 +1,81 @@
...
+                    />
+                    <OrientationInput
+                        className={styles.orientationInput}
+                        faramElementName="orientation"
+                        persistantHintAndError={false}
+                    />
+                    <TextInput
+                        type="number"
+                        // FIXME: string
+                        label="Font size"
```
Comment (tnagorra) [inline src/widgets/edit/Matrix2d/Row/DimensionContent/SubdimensionRow/index.js:42]: Use string

File: src/widgets/edit/Matrix2d/Row/DimensionContent/SubdimensionRow/index.js [hunk 23]
```diff
@@ -0,0 +1,81 @@
...
+                        className={styles.fontSizeInput}
+                        faramElementName="fontSize"
+                        persistantHintAndError={false}
+                    />
+                    <TextInput
+                        type="number"
+                        // FIXME: string
+                        label="Height"
```
Comment (tnagorra) [inline src/widgets/edit/Matrix2d/Row/DimensionContent/SubdimensionRow/index.js:50]: Use string

File: src/widgets/edit/Matrix2d/Row/DimensionContent/index.js [hunk 24]
```diff
@@ -0,0 +1,210 @@
...
+                                persistantHintAndError={false}
+                            />
+                            <OrientationInput
+                                className={styles.orientationInput}
+                                faramElementName="orientation"
+                                persistantHintAndError={false}
+                            />
+                            <TextInput
+                                type="number"
+                                label="Font size"
```
Comment (tnagorra) [inline src/widgets/edit/Matrix2d/Row/DimensionContent/index.js:124]: Use string

File: src/widgets/edit/Matrix2d/Row/DimensionContent/index.js [hunk 25]
```diff
@@ -0,0 +1,210 @@
...
+                                className={styles.fontSizeInput}
+                                faramElementName="fontSize"
+                                persistantHintAndError={false}
+                            />
+                            <TextInput
+                                type="number"
+                                label="Height"
```
Comment (tnagorra) [inline src/widgets/edit/Matrix2d/Row/DimensionContent/index.js:131]: Use string

File: src/widgets/edit/Matrix2d/Row/index.tsx [hunk 26]
```diff
@@ -0,0 +1,133 @@
...
+const emptyObject = {};
+
+export default class Row extends React.PureComponent<Props, State> {
+    private dimensionItemRendererParams = (key, elem, i) => ({
+        className: styles.dimensionContent,
+        data: elem,
+        faramElementName: String(i),
+        index: i,
+        keySelector: this.props.keySelector,
+        onEditButtonClick: () => {
```
Comment (tnagorra) [inline src/widgets/edit/Matrix2d/Row/index.tsx:43]: We shouldn't create new function inside renderer params

File: src/widgets/edit/Matrix2d/Row/index.tsx [hunk 27]
```diff
@@ -0,0 +1,133 @@
...
+            onGeoLinkModalVisiblityChange,
+            onLinkWidgetModalVisiblityChange,
+            onDimensionContentBackButtonClick,
+            selectedDimensionKey,
+            titleSelector,
+            widgetKey,
+        } = this.props;
+
+        const { dimensions = emptyList } = faramValues;
+        const selectedDimensionIndex = dimensions.findIndex(
```
Comment (tnagorra) [inline src/widgets/edit/Matrix2d/Row/index.tsx:64]: We should memoize expensive actions

**Pull request closed without merging**